src/documents/models.py

```diff
@@ -314,6 +397,83 @@ def thumbnail_file(self):
     def thumbnail_url(self):
         return reverse("fetch", kwargs={"kind": "thumb", "pk": self.pk})
 
+    def set_filename(self, filename):
+        if os.path.isfile(self.filename_to_path(filename)):
+            self.filename = filename
+
+
+def delete_empty_directory(directory):
+    if len(os.listdir(directory)) == 0:
+        try:
+            os.rmdir(directory)
+        except os.error:
+            # Directory not empty
+            pass
+
+
+@receiver(models.signals.m2m_changed, sender=Document.tags.through)
+@receiver(models.signals.post_save, sender=Document)
+def update_filename(sender, instance, **kwargs):
+    if instance.filename is None:
+        return
+
+    # Build the new filename
+    new_filename = instance.source_filename_new()
+
+    # If the filename is the same, then nothing needs to be done
+    if instance.filename is None or \
+       instance.filename == new_filename:
+        return
+
+    # Check if filename needs changing
+    if new_filename != instance.filename:
+        # Determine the full "target" path
+        path_new = instance.filename_to_path(new_filename)
+        dir_new = instance.filename_to_path(os.path.dirname(new_filename))
+
+        # Determine the full "current" path
+        path_current = instance.filename_to_path(instance.filename)
+
+        # Move file
+        try:
+            os.rename(path_current, path_new)
+        except PermissionError:
+            # Do not update filename in object
+            return
+
+        # Delete empty directory
+        old_dir = os.path.dirname(instance.filename)
+        old_path = instance.filename_to_path(old_dir)
+        delete_empty_directory(old_path)
+
+        instance.filename = new_filename
+
+        # Save instance
+        # This will not cause a cascade of post_save signals, as next time
+        # nothing needs to be renamed
+        instance.save()
+
+
+@receiver(models.signals.post_delete, sender=Document)
+def delete_files(sender, instance, **kwargs):
+    if instance.filename is None:
+        return
+
+    # Remove the document
+    old_file = instance.filename_to_path(instance.filename)
+
+    if os.path.isfile(old_file):
+        os.remove(old_file)
```

Comment:
There is a potential TOCTOU issue here: the file could be gone between when we check and when we remove. I'd instead suggest to `try` to call `os.remove()` and catch a `FileNotFoundError`, printing the warning message. If any other exception happens, we probably want to propagate it (as this code is doing already).

Reply:
done